README.md

```diff
@@ -127,40 +129,100 @@ $sslcert = @("EOT")
   account => "myaccount",
   authn_login => "host/myhost",
   authn_api_key => Sensitive("2z9mndg1950gcx1mcrs6w18bwnp028dqkmc34vj8gh2p500ny1qk8n"),
   ssl_certificate => $sslcert
-}]))
+}])
 ```
 
-#### Sensitive data type
+#### Example usage
 
-Note `conjur::secret` returns values wrapped in a `Sensitive` data type. In
-some contexts, such as string interpolation, it might cause surprising results
-(interpolating to `Sensitive [value redacted]`). This is intentional, as it
-makes it more difficult to accidentally mishandle secrets.
+```puppet
+node 'server-123' {
+  $db_password = Deferred(conjur::secret, ['inventory/db-password'])
+
+  file { '/tmp/creds.txt':
+    ensure => file,
+    mode => '0600',
+    content => $db_password,
+  }
+
+  file { '/tmp/creds.ini':
+    ensure => file,
+    mode => '0600',
+    content => Deferred('inline_epp', [
+      'password=<%= $db_password.unwrap %>',
+      { 'db_password' => $db_password }
+    ]),
+  }
+}
+```
 
-To use a secret as a string, you need to explicitly request it using the
-`unwrap` function; the result of the processing should be again wrapped in
-a `Sensitive` value.
+#### `Deferred` functions
 
-In particular, you should not pass unwrapped secrets as resource parameters
-if you can avoid it. Many resource types support `Sensitive` data type and
-handle it correctly. If a resource you're using does not, file a bug.
+This module _requires_ the use of Puppet v6+
+[`Deferred` functions](https://puppet.com/docs/puppet/6.17/deferring_functions.html)
+to ensure that the credential retrieval is fully handled on the agent. Since the
+resolution of variables is done on the agent _after_ the catalog is compiled, anything
+that requires the value of the variable during the compilation step (e.g. template
+compilation)
+[must also be wrapped in a `Deferred` invocation](https://puppet.com/docs/puppet/6.17/template_with_deferred_values.html).
+
+It is also important to note that you should make sure that you invoke the
+`conjur::secret` function using the proper `Deferred` syntax:
+
+```puppet
+# GOOD: Passing the parameters as an array
+Deferred(conjur::secret, ['production/postgres/password'])
+```
+
+```puppet
+# BAD: This will not work!
+Deferred(conjur::secret('production/postgres/password'))
+```
+
+You can read more about Puppet's `Deferred` functions
+[here](https://puppet.com/docs/puppet/6.17/deferring_functions.html).
+
+#### `Sensitive` type
+
+`conjur::secret` returns values wrapped in a `Sensitive` data type. In
+some contexts, such as string interpolation, it might cause surprising results
+(interpolating to `Sensitive [value redacted]`). This is intentional, as it
+makes it more difficult to accidentally mishandle secrets. Additionally, we also
+highly recommend that the whole `Deferred` invocation is also wrapped in
+`Sensitive` type to prevent accidental disclosure of information about the
+function parameters to the logs if more than just the API key is needed to be
+kept private.
```

Comment:
this is inconsistent with the rest of the docs - does this need to be updated?

Reply:
No - this is correct given the main usage doc changes. If only the API needs to be hidden from logs (which is true in majority of use cases), then the defaults are fine and `Sensitive` around the whole method is not needed. If things like `account`/`authn_login` need to be kept also protected, then we recommend that additional `Sensitive` is done on the whole `Deferred` call.

Reply:
if we "highly recommend" it, why do we not document it like this everywhere?

I think this could be clearer, if the use case is as you say.

Maybe on a separate line, we say:
> **Note**: We strongly recommend the API key from the Conjur configuration is marked `Sensitive`, but if any other data in the function parameters is also sensitive you may also wrap the whole `Deferred` invocation in the `Sensitive` type to prevent accidental disclosure of the sensitive information to the logs.

Is that a good representation of the recommendation? Do we also need to say, "Note that if you do this, you will need to `unwrap` the output twice."? Could we just recommend that the _parameters_ are marked `Sensitive` as well, or would that break our module (if another param is input as sensitive? (seems like a good troubleshooting case, tbh)